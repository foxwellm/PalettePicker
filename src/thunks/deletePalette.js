import { setLoading, setError, removePalette } from '../actions';
import { fetchData } from '../utils/api';

export const deletePalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/palettes/${id}`, 'DELETE');
      dispatch(removePalette(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}