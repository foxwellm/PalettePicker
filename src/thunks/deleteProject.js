import { setLoading, setError, removeProject } from '../actions';
import { fetchData } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/projects/${id}`, 'DELETE');
      dispatch(removeProject(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}