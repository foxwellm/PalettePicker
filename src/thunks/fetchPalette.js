import { setLoading, setError, setPalettes } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const palette = await fetchData(`/api/v1/palettes/${id}`, 'GET');
      dispatch(setPalettes([palette]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};