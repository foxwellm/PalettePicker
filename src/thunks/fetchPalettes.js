import { setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalettes = (allProjects) => {
  return (dispatch) => {
    const promisedPalettes = allProjects.map(async palette => {
      try {
        let allPalettes = await fetchData(`/api/v1/projects/${palette.id}/palettes`, 'GET');
        allPalettes.sort((a, b) => {
          if (a.updated_at > b.updated_at) return -1;
          if (b.updated_at < a.updated_at) return 1;
          return 0;
        });
        return allPalettes
      } catch (error) {
        dispatch(setError(error.message));
      }
    })
    return Promise.all(promisedPalettes);
  }
}