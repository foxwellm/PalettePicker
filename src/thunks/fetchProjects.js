import { setLoading, setError, setProjects } from '../actions';
