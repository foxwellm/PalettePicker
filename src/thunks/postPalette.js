import { setLoading, setError, addPalette } from '../actions';
import { fetchData } from '../utils/api';

export const postPalette = (palette) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const newPalette = await fetchData('/api/v1/palettes', 'POST', palette);
      dispatch(addPalette(newPalette[0]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false))
  }
}