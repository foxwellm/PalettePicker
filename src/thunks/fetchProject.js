import { setLoading, setError, setProjects, setPalettes } from '../actions';
import { fetchData } from '../utils/api';
import { fetchPalettes } from './fetchPalettes';

export const fetchProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const project = await fetchData(`/api/v1/projects/${id}`, 'GET');
      dispatch(setProjects([project]));
      const projectPalettes = await dispatch(fetchPalettes([project]));
      dispatch(setPalettes(projectPalettes));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};