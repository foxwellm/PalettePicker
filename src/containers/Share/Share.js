import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchProject } from '../../thunks/fetchProject';
import { fetchPalette } from '../../thunks/fetchPalette';
import ProjectContainer from '../ProjectContainer/ProjectContainer';
import Palette from '../Palette/Palette';
import NotFound from '../../components/NotFound/NotFound';
import PropTypes from 'prop-types';

export class Share extends Component {

  componentDidMount = () => {
    const { id, isProject, fetchProject, fetchPalette } = this.props;
    isProject ? fetchProject(id) : fetchPalette(id);
  };

  render() {
    const { projects, palettes, isProject, loading } = this.props;
    const palette = palettes[0];
    if(loading) {
      return <div>loading</div>
    } else if (isProject && projects.length) {
      return (
        <div>
          <ProjectContainer />
        </div>
      );
    } else if (!isProject && palettes.length) {
      return <Palette palette={palette} />
    } else {
      return <NotFound />
    }
  };
};

export const mapStateToProps = (state) => ({
  projects: state.projects,
  palettes: state.palettes,
  loading: state.loading,
});

export const mapDispatchToProps = (dispatch) => ({
  fetchProject: (id) => dispatch(fetchProject(id)),
  fetchPalette: (id) => dispatch(fetchPalette(id)),
});

Share.propTypes = {
  projects: PropTypes.array,
  palettes: PropTypes.array,
  fetchProject: PropTypes.func,
  fetchPalette: PropTypes.func,
}

Share.defaultProps = {
  palettes: [],
  projects: [],
}

export default connect(mapStateToProps, mapDispatchToProps)(Share);

