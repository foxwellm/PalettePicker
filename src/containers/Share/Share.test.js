import React from 'react';
import { Share, mapStateToProps, mapDispatchToProps } from './Share';
import { shallow } from 'enzyme';
import { fetchProject } from '../../thunks/fetchProject';
import { fetchPalette } from '../../thunks/fetchPalette';
jest.mock('../../thunks/fetchProject');
jest.mock('../../thunks/fetchPalette');

describe('Share', () => {
  let wrapper;
  let mockFetchProject;
  let mockFetchPalette;
  let mockProjects;
  let mockPalettes;
  let isProjectMock;

  beforeEach(() => {
    isProjectMock = true;
    mockFetchPalette = jest.fn();
    mockFetchProject = jest.fn();
    mockProjects = [{ id: 1, name: 'Project 1' }, { id: 2, name: 'Project 2' }];
    mockPalettes = [
      {
        id: 1,
        name: 'Palette 1',
        project_id: 1,
        color1: '#ffffff',
        color2: '#ffffff',
        color3: '#ffffff',
        color4: '#ffffff',
        color5: '#ffffff'
      },
      {
        id: 2,
        name: 'Palette 2',
        project_id: 1,
        color1: '#ffffff',
        color2: '#ffffff',
        color3: '#ffffff',
        color4: '#ffffff',
        color5: '#ffffff'
      }
    ];

    wrapper = shallow(
      <Share
        fetchProject={mockFetchProject}
        fetchPalette={mockFetchPalette}
        projects={mockProjects}
        palettes={mockPalettes}
        isProject={isProjectMock}
      />
    )
  });

  it('should match the correct snapshot for projects', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the correct snapshot for palettes', () => {
    isProjectMock = false;
    wrapper = shallow(
      <Share
        fetchProject={mockFetchProject}
        fetchPalette={mockFetchPalette}
        projects={mockProjects}
        palettes={mockPalettes}
        isProject={isProjectMock}
      />
    )
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the correct snapshot for not found', () => {
    isProjectMock = false;
    mockPalettes = [];
    wrapper = shallow(
      <Share
        fetchProject={mockFetchProject}
        fetchPalette={mockFetchPalette}
        projects={mockProjects}
        palettes={mockPalettes}
        isProject={isProjectMock}
      />
    )
    expect(wrapper).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('should call fetchProject', () => {
      wrapper.instance().componentDidMount();
      expect(mockFetchProject).toHaveBeenCalled();
    });

    it('should call fetchPalette', () => {
      isProjectMock = false;
      wrapper = shallow(
        <Share
          fetchProject={mockFetchProject}
          fetchPalette={mockFetchPalette}
          projects={mockProjects}
          palettes={mockPalettes}
          isProject={isProjectMock}
        />
      )
      wrapper.instance().componentDidMount();
      expect(mockFetchPalette).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object of props', () => {
      const mockState = {
        projects: mockProjects,
        palettes: mockPalettes,
        otherState: false,
      };
      const expected = {
        projects: mockProjects,
        palettes: mockPalettes,
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when fetchProject is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchProject();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchProject();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when fetchPalette is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchPalette();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchPalette();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

});