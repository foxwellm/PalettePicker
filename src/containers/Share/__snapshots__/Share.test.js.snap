// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share should match the correct snapshot for not found 1`] = `<NotFound />`;

exports[`Share should match the correct snapshot for palettes 1`] = `
<Connect(Palette)
  palette={
    Object {
      "color1": "#ffffff",
      "color2": "#ffffff",
      "color3": "#ffffff",
      "color4": "#ffffff",
      "color5": "#ffffff",
      "id": 1,
      "name": "Palette 1",
      "project_id": 1,
    }
  }
/>
`;

exports[`Share should match the correct snapshot for projects 1`] = `
<div>
  <Connect(ProjectContainer) />
</div>
`;
